379 abstract user model sausage hidden

DanielSeehausen opened the pull request:
closes #379 
closes #376 
closes #384 

I apologize for the size! There was a lot that was coupled and needed to be done to make this change. I popped everything out of staging and added selectively in a (hopefully) easier order to follow.

Reading through commit wise is a must if you do choose to read. Please see commits for more specific messages!
if you don't want to read through the whole thing, (which is totally understandable), please just leave a note what commits you were able to look at.

I will be click testing a lot of this tomorrow.

vdiezel commented on lib/ResourceAccess/Access.js:
```diff
@@ -0,0 +1,70 @@
+    return access.map(group => (
+      group.name === groupName
+        ? Group.addUser(group, userId)
+        : { ...group }
```
I think in functional programming it is okay to not copy the array elements that remain untouched - at least this was the pattern I have learned: 
https://redux.js.org/recipes/structuring-reducers/immutable-update-patterns

see section updating an item in an array: 
```js
function updateObjectInArray(array, action) {
  return array.map((item, index) => {
    if (index !== action.index) {
      // This isn't the item we care about - keep it as-is
      return item
    }

    // Otherwise, this is the one we want - return an updated value
    return {
      ...item,
      ...action.item
    }
  })
}
```

DanielSeehausen replied:
that makes sense - I will update now

DanielSeehausen commented:
just updated codeclimate to watch staging

vdiezel commented on lib/ResourceAccess/Group.js:
```diff
@@ -0,0 +1,44 @@
+  static get SCHEMA() {
+    return Joi.object({
+      name: Joi.string().min(1).max(45).regex(/^\S+$/).required(),
+      users: Joi.array().items(Joi.string(), Joi.number()).default([]),
```
why do we allow `Joi.number` here? userIds are always uuidv4

DanielSeehausen replied:
that's a good question
I was picturing access as a general thing, where the `users` just needs to be unique identifiers. If it were used with another resource that doesn't happen to use uuidv4, I wanted it to be compatible. I was trying to picture if this were an npm package, what it would need to be usable by anything with a `.access` on it

vdiezel replied:
YAGNI it :stuck_out_tongue:

DanielSeehausen replied:
YAGNI is always a good point

I want to push back on this because that would be making the module match its current use cases, instead of the module being true to what it offers in general. In this case, it offers to hold a collection of unique identifiers for users. Its pretty often those are numbers. While in this case we know all the ways it is being used (only with strings, for our purposes), it's not a guarantee all of its consumers use strings. I believe making it string would make it more brittle.

DanielSeehausen replied:
I should add I really do intend to make this an npm module

vdiezel commented on lib/ResourceAccess/Group.js:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  static addUser(group, userId) {
+    if (Group.userPresent(group, userId)) return { ...group }
```
also here we don't need to copy the group in the case the user is already present, just return as is (see comment in `Access` model.

DanielSeehausen replied:
updated

O-Susanoo commented on .eslintrc.js:
```diff
@@ -17,10 +18,7 @@ module.exports = {
-      "error",
-      { "allow": ["__MONGO_URI__", "__MONGO_DB_NAME__"] }
-    ],
+    "no-underscore-dangle": "off",
```
Like this one :+1:

DanielSeehausen replied:
that one goes out to you, @O-Susanoo

vdiezel commented on test/lib/ResourceAccess/Access.test.js:
```diff
@@ -0,0 +1,146 @@
+      expect(Access.userIsInGroup(access, userId, 'admins')).toBe(false)
+
+      access = Access.removeUserFromGroup(access, userId, 'operators')
+      expect(Access.userIsInGroup(access, userId, 'admins')).toBe(false)
```
here you want to check if a user is in operators I assume :)

DanielSeehausen replied:
very good eye thank you
PR review WORKS

DanielSeehausen replied:
updated

vdiezel commented on test/lib/ResourceAccess/Access.test.js:
```diff
@@ -0,0 +1,146 @@
+
+
+  describe('getUsersGroups', () => {
+    it('returns an array of the access groups a user in in', () => {
```
`in in`

DanielSeehausen replied:
not pedantic at all
good 👁️

DanielSeehausen commented:
hopefully just updated codecov to use staging as well

ConorTighe1995 commented on services/user/src/user/acceptDeviceInvitation.js:
```diff
@@ -1,26 +1,74 @@
+  // TODO: make whole thing a task
+  const deviceUsers = DeviceAccess.getUsers(device.access)
+
+  try {
```
Candidate for a function?

DanielSeehausen replied:
that's a good idea - I have a lot of controller stuff that is way longer than it should be. I updated - will go up in new commit ASAP

vdiezel commented on lib/ResourceAccess/Access.js:
```diff
@@ -0,0 +1,70 @@
+    return access.map(group => group.name)
+  }
+
+  static getUsersGroups(access, userId) {
```
With our current access rights definition it doesn't make sense to have a user in multiple groups, because the access rights are somewhat in a hierarchy. When we want to have users to be present on multiple groups, we need another access rights concepts that builds on composition (and `operator` would get only operating rights, but for viewing processes they would also need to be on the `viewer` group while now the `operator` group already includes view rights. 

We can keep the code as it is, but until we haven't finally designed authorization we should never have a user be present on multiple groups.

DanielSeehausen replied:
that's a good point, since ours is strictly a hierarchy. Maybe I should do one of these two things for now until we have further definition?:

These are both low effort options:
- ensure you can't add a user to a group if they are already found on the access object
- always switch the user out of group existing to group target if someone tries to add them to group target

What do you think @vdiezel ?

ConorTighe1995 replied:
If the access right are inspired by linux system permissions so I can understand why it might make sense from a design perspective as its also possible to create access groups on linux for different files and folders. But your right, from a user perspective we have this hierarchy system and then in the back-end it looks to me to be evolving into a resource orientated permissions system where each user might have very different rights to the next if they ever intersect groups.

ConorTighe1995 replied:
Will that happen here though? if not should this be `getUsersGroup`

DanielSeehausen replied:
so the second bullet point probably?

ConorTighe1995 replied:
@DanielSeehausen I didn't refresh my page so only seen your comment after I wrote that, yea I think that could reduce complexity and confusion in the future.

DanielSeehausen replied:
ill update it
meant to mention in the comments - would really like to make this access a key: value where:
```js
{ 
  viewers: { },
  admins: { },
}
```

at some point in the future. Hoping this is a step to standardize the interface to make that easier if it ever becomes worthwhile to do

vdiezel commented on lib/access-rights-lib.js:
```diff
@@ -196,18 +130,8 @@ function processAuthorizer(token){
-
-attachAuthorizer(RESOURCE.PROCESS, processAuthorizer);
-attachAuthorizer(RESOURCE.DEVICE, deviceAuthorizer);
+attachAuthorizer(RESOURCE.PROCESS, resourceAuthorizer)
+attachAuthorizer(RESOURCE.DEVICE, resourceAuthorizer)
```
:gun:

ConorTighe1995 commented on services/user/src/user/create.js:
```diff
@@ -19,22 +20,21 @@ async function create(event, context) {
   context.succeed(event) // triggers lambda to close via cognito request method
 
-  return success(result)
+  return success(newUser) // this return should not be doing anything if context.succeed is working as expected
```
If context fails or errors will this still return as usual? I had a quick look around for http-response in my IDE and the package.json of this repo and couldn't see it, does it have a failed method? would a flow like this possible:

```
try{
  context.succeed(event)
  return success(newUser)
}catch(e){
  return failed(e)
}
```

vdiezel replied:
We originally had this flow you describe here in all functions. We decided: If it fails, let it fail. Otherwise you can use `context.fail` in teh catch block.

vdiezel replied:
this handler needs some love and investigation anyway

DanielSeehausen replied:
I'm unsure. This was really me kicking the can down the road.

I don't believe that return can ever be hit (`context.succeed` ends execution as I understand), and on top of that the response isn't necessary because the caller doesn't make use of it. I just removed that line all together.

DanielSeehausen replied:
didn't see viktor's comment before I wrote
like he says. I removed the misleading return for now

vdiezel commented on lib/models/User/Notifications/Notification.js:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  static get NOTIFICATION_TOPICS() {
+    return [ 'device/invitation', 'changelog' ]
```
we also need `device/revocation` as a topic here, otherwise our device access revocations will not work!

DanielSeehausen replied:
added it!

vdiezel commented on lib/models/User/Notifications/Notification.js:
```diff
@@ -0,0 +1,45 @@
+      from: Joi.object({
+        type: Joi.string().required(),
+        name: Joi.string().required(),
+        originId: [ Joi.number(), Joi.string() ],
```
we miss the keys `sourceName` and `sourceType` here - they are being destructured here: 
https://github.com/ospin-web-dev/hambda/blob/2ac7007c46e96f19398727df3ccc2b4ee9f0f49d/services/user/src/user/patchAllNotifications.js#L9

but actually never used. The vales they curently have are those: 
https://github.com/ospin-web-dev/REACTor/blob/master/.circleci/publishChanges.js#L35

DanielSeehausen replied:
as I understand it, those keys are serialized in the controller:
```
 // patchAllNotifications.js
  const notification = Notification.new({
    topic,
    type: 'persistent',
    from: {
      type: sourceType,
      name: sourceName,
      originId,
    },
    data: { message },
  })
```

DanielSeehausen replied:
and in the DB:
![image](https://user-images.githubusercontent.com/17087405/80087451-628ac080-855b-11ea-9f94-f3aa02c70cf3.png)

DanielSeehausen replied:
tests also seem to be using type/topic, so this is strange in general...might need to look into this more

vdiezel commented on lib/models/User/Notifications/Notification.js:
```diff
@@ -0,0 +1,45 @@
+        deviceId: Joi.string().pattern(UUIDV4_REGEX), // Only on invitations TODO: this needs to be made a sub schema when device model is compelted
+        message: Joi.string().default(''),
+      }).required(),
+      read: Joi.bool(), // TODO: is this still required?
```
I think so - this should be used e.g. for device invitations: they are persistent, but as long as they are unread, the notifications icon on the side bar in REACTor remains red. Once read, the icon turns grey. Did this to prevent annoying users when they have persistent messages.

DanielSeehausen replied:
👍 noice

ConorTighe1995 commented on services/user/src/user/getChangelog.js:
```diff
@@ -1,14 +1,19 @@
 exports.main = HTTPDefaultMiddleware(getChangelog, {
   useDoNotWaitForEmptyEventLoop: false,
-  useHTTPEventNormalizer: false, // why?
+  useHTTPEventNormalizer: false, // why? -VD || 'I dont know either' -DS
```
Do you know the after effects of setting this to false? How will how will it normalizer the data?

DanielSeehausen replied:
the event normalizer just serializes the event for value access convenience, IIRC

ConorTighe1995 replied:
Ok just curious because of the comments

ConorTighe1995 commented on services/user/src/user/revokeIotPolicy.js:
```diff
@@ -1,27 +1,60 @@
+  ])
 
-  validateRevokeIotPolicy(requestData)
+  await IotAPI.detachDevicePolicyFromUser(revocation.data.deviceId, cognitoIdentityId)
```
Maybe my eyes are playing tricks on me, but you should switch line 52 and 54 as revocation is being hoisted

vdiezel replied:
great catch - that is my bad, I made Daniel do this (because logically it makes sense). We need the `deviceId` which is stored on the revocation in the DB. Guess we have to fetch it first between lines 50 and 52.

vdiezel replied:
I also see that the revocation notifications is actually never removed from the user here, or?

DanielSeehausen replied:
very good catch - even the linter was complaining and I entirely did not see it.
I just reversed them. 

it shouldn't hoist in that scenario, so the code was definitely broken. At the top of the controller I mentioned this needs a 200 test...I guess the fact that such a big issue wasn't caught by a test means I should add that now.

DanielSeehausen replied:
added tests

ConorTighe1995 replied:
Ahh yes hoisting is var only, good point

ConorTighe1995 commented on lib/tasks/users/destroyUser.js:
```diff
@@ -0,0 +1,40 @@
+
+
+module.exports = async userId => {
+  /* We need to do the following cleanup:
```
Do theses comments still need to be here now that the below code is in place?

DanielSeehausen replied:
no! I removed them

ConorTighe1995 commented on services/user/src/user/delete.js:
```diff
@@ -1,22 +1,44 @@
+  ])
 
-  const result = await user.delete(requestData)
+  // Note: this is more than removing the user from the db. See the task
```
If this requires a note should then name be more descriptive? something like removeUserFromSystem or deleteAllUserRelatedData

DanielSeehausen replied:
good idea - i will go for `deleteAllUserRelatedData`

DanielSeehausen replied:
updated

ConorTighe1995 commented on services/user/test/user/delete.test.js:
```diff
@@ -6,8 +6,6 @@ const Process = require('models/Process/Process')
 const Seeder = require('test-helpers/seeder/Seeder')
 const mongo = require('mongodb-lib')
 const CognitoAPI = require ('cognito/CognitoAPI')
-//process.env.AWS_COGNITO_USERPOOL_ID = 'MERKEL!'
```
lol

DanielSeehausen replied:
:))

vdiezel commented on lib/tasks/users/destroyUser.js:
```diff
@@ -0,0 +1,40 @@
+  */
+  const user = await User.find(userId)
+  await Promise.all([
+    removeUserFromDevices(user),
+    removeUserFromProcesses(user),
```
Nesting `Promise.all` - so easy!

DanielSeehausen replied:
glad you like it

vdiezel commented on services/user/src/user/patchAllNotifications.js:
```diff
@@ -1,26 +1,46 @@
+    },
+  } = event
+
+  // I believe any user could inject the correct sourceName/Type values and fire this! This endpoint may be lacking some security
```
No, this endpoint is not exposed to the public via HTTP but is called via the CD suite which has an AWS profile sufficient enough to envoke this using the `aws-sdk`. You can see here, is has no triggers: 
https://github.com/ospin-web-dev/hambda/blob/staging/services/user/serverless.yml#L368. So we are safe here and this comment can be removed :)

DanielSeehausen replied:
nice - thanks for the education. removing comment

ConorTighe1995 commented on lib/models/User/User.js:
```diff
@@ -0,0 +1,108 @@
+const Connectable = require('models/Connectable')
+const UserNotifications = require('models/User/Notifications/UserNotifications')
+
+const UUIDV4_REGEX = /[a-f0-9]{8}(-[a-f0-9]{4}){3}-[a-f0-9]{12}/
```
This appears in a few places, might be worth introducing a regex utility file where we can update this easily in the future and expand on regex functions. would DRY out the code and save time when validating and testing regex.

DanielSeehausen replied:
yes I agree - have been lazy for quite some time just always copying this.

While tempting to monkey patch String, I'll instead add it to a util or something :)

DanielSeehausen replied:
i added this!

DanielSeehausen commented:
thank you guys very much for fast review - I updated and made all change requests except for one (re: access unique identifiers - I left a comment)

The biggest changes are to Access. it is now implemented to only allow a single user. See tests for documentation on that
